src/Semantic/Git.hs

```diff
@@ -32,24 +41,29 @@ clone url path = sh $ do
 -- | git cat-file -p
 catFile :: FilePath -> OID -> IO Text
 catFile gitDir (OID oid) = sh $ do
-  run "git" ["-C", pack gitDir, "cat-file", "-p", oid]
+  run "git" ["-C", pack gitDir, "cat-file", "-p", Text.decodeUtf8 oid]
 
 -- | git ls-tree -rz
 lsTree :: FilePath -> OID -> IO [TreeEntry]
-lsTree gitDir (OID sha) = sh $ parseEntries <$> run "git" ["-C", pack gitDir, "ls-tree", "-rz", sha]
+lsTree gitDir (OID sha) = Streaming.Process.withStreamingOutput lsproc go
+  where
+    lsproc = Streaming.Process.proc "git" ["-C", gitDir, "ls-tree", "-zr", BC.unpack sha]
+    go :: ByteStream.ByteString IO () -> IO [TreeEntry]
+    go = fmap (fromRight [] . fst) . AP.Stream.parse everything
 
 sh :: MonadIO m => Sh a -> m a
 sh = shelly . silently . onCommandHandles (initOutputHandles (`hSetBinaryMode` True))
 
 -- | Parses an list of entries separated by \NUL, and on failure return []
-parseEntries :: Text -> [TreeEntry]
+parseEntries :: BC.ByteString -> [TreeEntry]
 parseEntries = fromRight [] . AP.parseOnly everything
-  where
-    everything = AP.sepBy entryParser "\NUL" <* "\NUL\n" <* AP.endOfInput
+
+everything :: Parser [TreeEntry]
+everything = AP.sepBy entryParser "\NUL"
```

Comment:
This looks to parse differently: only entries separated by `\NUL`, no trailing `\NUL` & newline, and no `AP.endOfInput`. Is this correct?

Reply:
The removal of the trailing newline is correct; that was inserted by `shelly`. For paranoia’s sake, I’ll insert the `endOfInput` calls back. Ultimately it doesn’t matter much, since this is all after we’ve parsed the last entry.